staging/src/k8s.io/client-go/restmapper/shortcut_test.go

```diff
@@ -255,6 +259,148 @@ func TestKindForWithNewCRDs(t *testing.T) {
+							Name:       "horizontalpodautoscalers",
+							ShortNames: []string{"hpa"},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:        "warn-builtin-shortname-ambugity",
+			arg:         "po",
+			expected:    schema.GroupVersionResource{Resource: "pods", Group: ""},
+			expectedLog: `short name "po" could also match lower priority resource poddlers.acme.com`,
+			srvRes: []*metav1.APIResourceList{
+				{
+					GroupVersion: "v1",
+					APIResources: []metav1.APIResource{{Name: "pods", SingularName: "pod", ShortNames: []string{"po"}}},
+				},
+				{
+					GroupVersion: "acme.com/v1",
+					APIResources: []metav1.APIResource{{Name: "poddlers", ShortNames: []string{"po"}}},
+				},
+			},
+		},
+		{
+			name:        "warn-builtin-shortname-ambugity-multi-version",
+			arg:         "po",
+			expected:    schema.GroupVersionResource{Resource: "pods", Group: ""},
+			expectedLog: `short name "po" could also match lower priority resource poddlers.acme.com`,
+			srvRes: []*metav1.APIResourceList{
+				{
+					GroupVersion: "v1",
+					APIResources: []metav1.APIResource{{Name: "pods", SingularName: "pod", ShortNames: []string{"po"}}},
+				},
+				{
+					GroupVersion: "acme.com/v1",
+					APIResources: []metav1.APIResource{{Name: "poddlers", ShortNames: []string{"po"}}},
+				},
+				{
+					GroupVersion: "acme.com/v1beta1",
+					APIResources: []metav1.APIResource{{Name: "poddlers", ShortNames: []string{"po"}}},
+				},
+			},
+		},
+		{
+			name:     "resource-match-singular-preferred",
+			arg:      "pod",
+			expected: schema.GroupVersionResource{Resource: "pod", Group: ""},
+			srvRes: []*metav1.APIResourceList{
+				{
+					GroupVersion: "v1",
+					APIResources: []metav1.APIResource{{Name: "pods", SingularName: "pod"}},
+				},
+				{
+					GroupVersion: "acme.com/v1",
+					APIResources: []metav1.APIResource{{Name: "poddlers", ShortNames: []string{"pods", "pod"}}},
+				},
+			},
+		},
+		{
+			name:        "resource-multiple-versions-shortform",
+			arg:         "hpa",
+			expected:    schema.GroupVersionResource{Resource: "horizontalpodautoscalers", Group: "autoscaling"},
+			expectedLog: "",
+			srvRes: []*metav1.APIResourceList{
+				{
+					GroupVersion: "autoscaling/v1alphav1",
+					APIResources: []metav1.APIResource{
+						{
+							Name:       "horizontalpodautoscalers",
+							ShortNames: []string{"hpa"},
+						},
+					},
+				},
+				{
+					GroupVersion: "autoscaling/v1",
+					APIResources: []metav1.APIResource{
+						{
+							Name:       "horizontalpodautoscalers",
+							ShortNames: []string{"hpa"},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		ds := &fakeDiscoveryClient{}
+		ds.serverResourcesHandler = func() ([]*metav1.APIResourceList, error) {
+			return test.srvRes, nil
+		}
+		warnedCount := 0
+		mapper := NewShortcutExpander(&fakeRESTMapper{}, ds, func(a string) {
+			warnedCount += 1
+			if warnedCount > 1 {
+				t.Fatalf("warning fired more than expected")
+			}
```

Comment:
I think we actually want a testcase that exercises multiple warnings since that's a scenario this PR introduces

Reply:
you could append to a list of warnings encountered, then compare the list received to a list of warnings expected

Reply:
I've updated test function to handle multiple warnings. Let me know what you think?. Thanks.

Reply:
@liggitt would you mind taking a look when you have a chance?. Thanks.